src/domain/services/jwt.subscription.service.ts

```diff
@@ -0,0 +1,133 @@
+import { Frequency } from '../entities/subscription.entity.js';
+import { EmailProvider } from 'src/domain/ports/secondary/email.provider.js';
+import { TemplateProvider } from '../ports/secondary/templates.provider.js';
+import { Logger, LoggerProvider } from '../ports/secondary/logger.provider.js';
+import { Weather } from '../entities/weather.entity.js';
+import { WeatherService } from '../ports/primary/weather.service.js';
+import { SubscriptionService } from '../ports/primary/subscription.service.js';
+
+export type SubscribeOptions = {
+  email: string;
+  city: string;
+  frequency: Frequency;
+};
+
+type Options = {
+  jwtProvider: JwtProvider;
+  subscriptionRepository: SubscriptionRepository;
+  weatherService: WeatherService;
+  emailProvider: EmailProvider;
+  templateProvider: TemplateProvider;
+  loggerProvider: LoggerProvider;
+  config: { APP_URL: string };
+};
+
+export class JwtSubscriptionService implements SubscriptionService {
+  private jwtProvider: JwtProvider;
+  private subscriptionRepository: SubscriptionRepository;
+  private weatherService: WeatherService;
+  private emailProvider: EmailProvider;
+  private templateProvider: TemplateProvider;
+  private logger: Logger;
+
+  private appUrl: string;
+
+  constructor({
+    jwtProvider,
+    subscriptionRepository,
+    weatherService,
+    emailProvider,
+    templateProvider,
+    loggerProvider,
+    config,
+  }: Options) {
+    this.jwtProvider = jwtProvider;
+    this.subscriptionRepository = subscriptionRepository;
+    this.weatherService = weatherService;
+    this.emailProvider = emailProvider;
+    this.templateProvider = templateProvider;
+
+    this.logger = loggerProvider.createLogger('SubscriptionService');
+
+    this.appUrl = config.APP_URL;
+  }
+
+  private makeUnsubscribeLink(subscriptionId: string) {
+    return `${this.appUrl}/api/unsubscribe/${subscriptionId}`;
+  }
+
+  public async handleSubscriptions(frequency: Frequency) {
+    this.logger.info({ msg: 'Handling weather subscription has been started', frequency });
+
+    const dataloader = new Dataloader<string, Weather>(async (cities) => {
+      return await Promise.all(
+        cities.map(async (city) => {
+          return await this.weatherService.getWeather(city);
+        }),
+      );
+    });
+
+    for await (const subscriptions of this.subscriptionRepository.iterateSubscriptions(frequency)) {
+      for (const { id, email, city } of subscriptions) {
+        const weather = await dataloader.load(city);
+        const unsubscribeLink = this.makeUnsubscribeLink(id);
+
+        await this.emailProvider.sendEmail({
+          to: [email],
+          template: this.templateProvider.getWeatherUpdateTemplate({
+            city,
+            unsubscribeLink,
+            ...weather,
+          }),
+        });
+      }
+    }
+
+    this.logger.info({ msg: 'Handling weather subscription has been finished', frequency });
+  }
+
+  private async assertIsSubscriptionExists(email: string) {
```

Comment:
"exists" is better than "is exists"
```suggestion
  private async assertSubscriptionExists(email: string) {
```

Reply:
Done https://github.com/GenesisEducationKyiv/software-engineering-school-5-0-allohamora/pull/8/commits/8d8453277ec90b5ba5a9f083a1cf95b679ccd5ed